File: CS/dxExampleE859(CS)/WinSolution.Win/WinApplication.cs
using System;
using DevExpress.ExpressApp.Win;
using DevExpress.ExpressApp;
using Northwind;
using DevExpress.Xpo;
using System.Linq;
using Dennis.Linq;

namespace WinSolution.Win {
    public partial class WinSolutionWindowsFormsApplication : WinApplication {
        public WinSolutionWindowsFormsApplication() {
            InitializeComponent();
        }
        private void WinSolutionWindowsFormsApplication_DatabaseVersionMismatch(object sender, DevExpress.ExpressApp.DatabaseVersionMismatchEventArgs e) {
            e.Updater.Update();
            e.Handled = true;
        }
        #region Linq Support in XAF
        protected override CollectionSourceBase CreateCollectionSourceCore(DevExpress.ExpressApp.ObjectSpace objectSpace, Type objectType, string listViewID) {
            CollectionSourceBase cs = null;
            switch (listViewID) {
                case "Customer_ListView_Linq":
                    XPQuery<Customer> customers = new XPQuery<Customer>(objectSpace.Session);
                    var queryCustomers = from c in customers
                                         where (c.Country == "Germany" && c.ContactTitle == "Sales Representative")
                                         orderby c.ContactName
                                         select c;
                    cs = new LinqCollectionSource(objectSpace, objectType, queryCustomers);
                    break;
                case "Order_ListView_Linq":
                    XPQuery<Order> orders = new XPQuery<Order>(objectSpace.Session);
                    var queryOrders = from o in orders
                                      orderby o.Employee.FirstName ascending
                                      group o by o.Employee.FirstName + " " + o.Employee.LastName into oo
                                      where oo.Sum(o => o.Freight) > 10000
                                      select new
                                      {
                                          Employee_Linq = oo.Key,
                                          Orders_Sum_Linq = oo.Sum(o => o.Freight)
                                      };
                    cs = new LinqCollectionSource(objectSpace, objectType, queryOrders);
                    break;
                case "Employee_ListView_Linq":
                    XPQuery<Employee> employees = new XPQuery<Employee>(objectSpace.Session);
                    var queryEmployees = from e in employees
                                         select new
                                         {
                                             FullName = e.FirstName + " " + e.LastName,
                                             Orders_Max_Linq = e.Orders.Max(o => o.Freight)
                                         };

                    cs = new LinqCollectionSource(objectSpace, objectType, queryEmployees);
                    break;
                default:
                    cs = base.CreateCollectionSourceCore(objectSpace, objectType, listViewID);
                    break;
            }
            return cs;
        }
        #endregion
    }
}

File: VB/WinSolution.Win/WinApplication.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.ExpressApp.Win
Imports DevExpress.ExpressApp
Imports Northwind
Imports DevExpress.Xpo
Imports System.Linq
Imports Dennis.Linq

Namespace WinSolution.Win
	Partial Public Class WinSolutionWindowsFormsApplication
		Inherits WinApplication
		Public Sub New()
			InitializeComponent()
		End Sub
		Private Sub WinSolutionWindowsFormsApplication_DatabaseVersionMismatch(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.DatabaseVersionMismatchEventArgs) Handles MyBase.DatabaseVersionMismatch
			e.Updater.Update()
			e.Handled = True
		End Sub
		#Region "Linq Support in XAF"
		Protected Overrides Function CreateCollectionSourceCore(ByVal objectSpace As DevExpress.ExpressApp.ObjectSpace, ByVal objectType As Type, ByVal listViewID As String) As CollectionSourceBase
			Dim cs As CollectionSourceBase = Nothing
			Select Case listViewID
				Case "Customer_ListView_Linq"
					Dim customers As XPQuery(Of Customer) = New XPQuery(Of Customer)(objectSpace.Session)
					Dim queryCustomers = From c In customers _
					                     Where (c.Country = "Germany" AndAlso c.ContactTitle = "Sales Representative") _
					                     Order By c.ContactName _
					                     Select c
					cs = New LinqCollectionSource(objectSpace, objectType, queryCustomers)
				Case "Order_ListView_Linq"
					Dim orders As XPQuery(Of Order) = New XPQuery(Of Order)(objectSpace.Session)
                    Dim queryOrders = From o In orders _
                                      Order By o.Employee.FirstName Ascending _
                                      Group o By o.Employee.FirstName Into oo = Group _
                                      Where oo.Sum(Function(o) o.Freight) > 10000 _
                                      Select New With {Key .Employee_Linq = FirstName, Key .Orders_Sum_Linq = oo.Sum(Function(o) o.Freight)}
                    cs = New LinqCollectionSource(objectSpace, objectType, CType(queryOrders, IQueryable))
				Case "Employee_ListView_Linq"
					Dim employees As XPQuery(Of Employee) = New XPQuery(Of Employee)(objectSpace.Session)
                    Dim queryEmployees = From e In employees _
                                         Select New With {e.FirstName, Key .Orders_Max_Linq = e.Orders.Max(Function(o) o.Freight)}

					cs = New LinqCollectionSource(objectSpace, objectType, queryEmployees)
				Case Else
					cs = MyBase.CreateCollectionSourceCore(objectSpace, objectType, listViewID)
			End Select
			Return cs
		End Function
		#End Region
	End Class
End Namespace
